docs/spec_draft.md

```diff
@@ -1940,6 +1941,40 @@ produces a `result` tensor. For boolean tensors, it computes the logical NOT.
 
 [Back to Ops](#index-of-ops)
 
+
+## stablehlo.optimization_barrier
+
+### Semantics
+
+Blocks any optimization pass from moving computations across the barrier. It
+ensures that the operations producing the `operand` are executed before any
+operations that depend on `result`.
+
```

Comment:
Ensures that the operations that produce the `operand` are executed before any
operations that depend on the `result`, preventing any optimisation pass from
moving computations across the barrier.

Reply:
Thanks for the suggestion! I was also trying to merge the two sentence like above but following is what bothers me:

It gives the impression that ensuring `operations that produce the operand are executed before any operations that depend on the result` is sufficient to `preventing any optimization pass from ....`. 

IMO this op is also a directive to the compiler  to  prevent any optimization which needs moving computations around the barriers even if those computations are not producers or consumers of the `optimization_barrier` op.  

Comparing with  [AfterAll](https://github.com/openxla/stablehlo/blob/main/docs/spec_draft.md#stablehloafter_all) which has similar wordings but, IMO, it can allow reordering computations across the op if those computations are not producers or consumers of the `afterAll` op. 

how about using
```
Ensures that the operations that produce the `operand` are executed before any
operations that depend on the `result` and prevents any optimization pass from
moving computations across the barrier.
```

Reply:
**upd** after some exploration: 

`AfterAll`, with input /output being tokens, can only be used to introduce data dependency between side-effecting operations (infeed/outfeed/send//recv). Where as `optimization_barrier` can be used to introduce the same between any (sideeffecting or not) operations. Other than that the two operations are the same. 

IMO, the line `preventing any optimization pass from moving computations across the barrier.` specifically means the optimizations like CSE that involves non-side-efecting operations and hence the same line in not present in AfterAll case.

Also, I believe both the operations can move operations across them of they are not the consumers or producers of the op.

Having said that I am tempted to adopt your suggestion with a slight modification:
```
Ensures that the operations that produce the `operand` are executed before any
operations that depend on the `result`, preventing any optimisation pass from
moving computations, which are producers or consumers of the `optimization_barrier` operation, across the barrier.
```

Reply:
`Also, I believe both the operations allow the opt passes to move operations across them if those operations are not the consumers or producers of the op. And I think introducing this detail whould complicate the spec.`

In this case `prevents any optimisation pass from moving computations across the barrier` is no longer true for either op.

Reply:
I agree that this part of the sentence is "loose" in its intent.  Maybe refactoring it to the following will help.
... 
prevents any optimization pass from moving computations, which are either the producers or consumers of the optimization_barrier op, across the barrier. 
```

Let us seek opinion from the next reviewer.